src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -466,7 +524,44 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
             if (NameResolutionPal.SupportsGetAddrInfoAsync && ipAddress is null)
             {
                 ValidateHostName(hostName);
-                return NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+
+                if (NameResolutionTelemetry.IsEnabled)
+                {
+                    ValueStopwatch stopwatch = NameResolutionTelemetry.Log.ResolutionStart(hostName);
+
+                    Task coreTask;
+                    try
+                    {
+                        coreTask = NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+                    }
+                    catch when (LogFailure(hostName, stopwatch))
+                    {
+                        Debug.Fail("LogFailure should return false");
+                        throw;
+                    }
+
+                    coreTask.ContinueWith(
+                        (task, state) =>
+                        {
+                            var tuple = (Tuple<string, ValueStopwatch>)state!;
+                            NameResolutionTelemetry.Log.AfterResolution(
+                                hostNameOrAddress: tuple.Item1,
+                                stopwatch: tuple.Item2,
+                                successful: task.IsCompletedSuccessfully);
+                        },
+                        state: Tuple.Create(hostName, stopwatch),
```

Comment:
This allocates, just use async/await instead.

Reply:
See [comment](https://github.com/dotnet/runtime/pull/38409/files#diff-434fc33b2da88cc8b47b46b328d07222R557-R559) - the task from `NameResolutionPal.GetAddrInfoAsync` has to be returned directly.

Reply:
Marking this as resolved as I do not believe it is possible to implement this with async await since we need control over the returned Task.

As the logic now also checks that the event will be fired (`stopwatch.IsActive`) before attaching the continuation, an async await approach would also have more overhead in the scenario where only counters are enabled.